aws/aws-sdk-js
# Support IPv6 endpoints in IMDS

Adds support for IMDS IPv6 endpoints

### Manual testing
Testing was done by printing endpoint as follows:

```diff
--- a/lib/metadata_service.js
+++ b/lib/metadata_service.js
@@ -60,6 +60,7 @@ AWS.MetadataService = inherit({
    *   retry delay on retryable errors. See AWS.Config for details.
    */
   constructor: function MetadataService(options) {
+    console.log({ endpoint: this.endpoint });
     if (options && options.host) {
       options.endpoint = 'http://' + options.host;
       delete options.host;
```

<details>
<summary>test-v2.js</summary>

```js
const AWS = require("./aws-sdk-js/index.js");
  
const testCredentials = async () => {
  const region = "us-west-2";
  const client = new AWS.DynamoDB({ region });
  await client.listTables().promise();
};

module.exports = {
  testCredentials,
};
```

</details>

<details>
<summary>test-v2.shared.js</summary>

```js
import { promisify } from "util";
import { exec } from "child_process";

const execPromise = promisify(exec);

const testCredentialsString = `const { testCredentials } = require("./test-v2.js"); (async() => {await testCredentials();})();`;

export const test = async (envKeyValuePair) => {
  console.log(`ENV: ${envKeyValuePair}`);
  const command = `${envKeyValuePair ? envKeyValuePair : ""} node -e '${testCredentialsString}'`;
  try {
    const { stdout } = await execPromise(command);
    console.log(stdout);
  } catch (error) {
    console.log(error.stderr.split("\n").filter(line => line.includes("Error")));
  }
};
```

</details>

#### Environment variables

<details>
<summary>test-v2.env.mjs</summary>

```js
import { test } from "./test-v2.shared.mjs";

const ENV_ENDPOINT_NAME = "AWS_EC2_METADATA_SERVICE_ENDPOINT";
const ENV_ENDPOINT_MODE_NAME = "AWS_EC2_METADATA_SERVICE_ENDPOINT_MODE";

await test();
const endpoints = [
  "http://169.254.169.254",
  "http://[fd00:ec2::254]",
  "invalidEndpoint",
];
for (const endpoint of endpoints) {
  await test(`${ENV_ENDPOINT_NAME}=${endpoint}`);
}

await test();
const endpointModes = ["IPv4", "IPv6", "invalidEndpointMode"];
for (const endpointMode of endpointModes) {
  await test(`${ENV_ENDPOINT_MODE_NAME}=${endpointMode}`);
}
```

</details>

<details>
<summary>Output</summary>

```console
ENV: undefined
{ endpoint: 'http://169.254.169.254' }

ENV: AWS_EC2_METADATA_SERVICE_ENDPOINT=http://169.254.169.254
{ endpoint: 'http://169.254.169.254' }

ENV: AWS_EC2_METADATA_SERVICE_ENDPOINT=http://[fd00:ec2::254]
{ endpoint: 'http://[fd00:ec2::254]' }

ENV: AWS_EC2_METADATA_SERVICE_ENDPOINT=invalidEndpoint
[
  'TypeError [ERR_INVALID_URL]: Invalid URL: invalidEndpoint',
  '    at onParseError (internal/url.js:279:9)'
]
ENV: undefined
{ endpoint: 'http://169.254.169.254' }

ENV: AWS_EC2_METADATA_SERVICE_ENDPOINT_MODE=IPv4
{ endpoint: 'http://169.254.169.254' }

ENV: AWS_EC2_METADATA_SERVICE_ENDPOINT_MODE=IPv6
{ endpoint: 'http://[fd00:ec2::254]' }

ENV: AWS_EC2_METADATA_SERVICE_ENDPOINT_MODE=invalidEndpointMode
[
  "      throw new Error('Unsupported endpoint mode: ' + endpointMode);",
  'Error: Unsupported endpoint mode: invalidEndpointMode'
]
```

</details>

#### Shared ini configuration

<details>
<summary>Config</summary>

```console
[profile imds_mode_v4]
ec2_metadata_service_endpoint_mode=IPv4

[profile imds_mode_v6]
ec2_metadata_service_endpoint_mode=IPv6

[profile imds_mode_invalid]
ec2_metadata_service_endpoint_mode=invalid

[profile imds_endpoint_v4]
ec2_metadata_service_endpoint=http://169.254.169.254

[profile imds_endpoint_v6]
ec2_metadata_service_endpoint=http://[fd00:ec2::254]

[profile imds_endpoint_invalid]
ec2_metadata_service_endpoint=invalid
```

</details>

<details>
<summary>test-v2.config.mjs</summary>

```js
import { test } from "./test-v2.shared.mjs";
  
await test();
const endpointConfigs = [
  "imds_mode_v4",
  "imds_mode_v6",
  "imds_mode_invalid",
];
for (const endpoint of endpointConfigs) {
  await test(`AWS_PROFILE=${endpoint}`);
}

await test();
const endpointModeConfigs = ["imds_endpoint_v4", "imds_endpoint_v6", "imds_endpoint_invalid"];
for (const endpointMode of endpointModeConfigs) {
  await test(`AWS_PROFILE=${endpointMode}`);
}
```

</details>

<details>
<summary>Output</summary>

```console
ENV: undefined
{ endpoint: 'http://169.254.169.254' }

ENV: AWS_PROFILE=imds_mode_v4
{ endpoint: 'http://169.254.169.254' }

ENV: AWS_PROFILE=imds_mode_v6
{ endpoint: 'http://[fd00:ec2::254]' }

ENV: AWS_PROFILE=imds_mode_invalid
[
  "      throw new Error('Unsupported endpoint mode: ' + endpointMode);",
  'Error: Unsupported endpoint mode: invalid'
]
ENV: undefined
{ endpoint: 'http://169.254.169.254' }

ENV: AWS_PROFILE=imds_endpoint_v4
{ endpoint: 'http://169.254.169.254' }

ENV: AWS_PROFILE=imds_endpoint_v6
{ endpoint: 'http://[fd00:ec2::254]' }

ENV: AWS_PROFILE=imds_endpoint_invalid
[
  'TypeError [ERR_INVALID_URL]: Invalid URL: invalid',
  '    at onParseError (internal/url.js:279:9)'
]
```

</details>

##### Checklist


- [x] changelog is added, `npm run add-change`
